Parser improvements

vogelsgesang opened the pull request:
I am currently getting started with MessageFormat and as a starting point I was reading your definition of the MessageFormat grammar.

Doing so, I noticed some minor things which could be done to simplify the pegjs grammar:

* PegJS grammar:
    * Remove `messageFormatPatternRight` and directly use `messageFormatElement` within `messageFormatPattern`
    * Remove `pluralStyle` and `selectStyle` which were just forwarding to `pluralFormatPattern` respectively `selectFormatPattern`
    * simplified parsing of strings and ids
* automatically compile the grammar in the development version
* I added two new testcases for trailing and leading whitespace in strings

All testcases are passing for messageformat.js and lib/messageformat.dev.js

SlexAxton commented:
It looks like there are a bunch of failing tests in Travis, are those not from this change set?

vogelsgesang commented:
Well, the Travis CI tests are not passing since I wrote them for the new grammar and the new syntax tree. For my local tests I replaced `test/common.js` with:

```
MessageFormat = require('../lib/messageformat.dev');
```

instead of

```
MessageFormat = require('../messageformat');
```

But I did not want to commit this change since it might break your testing infrastructure.
What should I do about this?

vogelsgesang commented:
Or should I apply the changes I applied to `lib/messageformat.dev.js` to `messageformat.js` as well?
Sorry, I guess I just didn't understand your workflow

vogelsgesang commented:
Ok, after having a look at some other pull-requests I noticed that changes have to be applied to both `messageformat.js` and `lib/messageformat.dev.js`. So I copied my changes over.

The tests in Travis CI on nodejs version 0.10 are passing now. The failing tests under nodejs 0.8 were not from this changeset and are already failing on your current master branch.

SlexAxton commented:
`messageformat.js` is a generated file. It's mostly a wrapper and adds in the parser. If my memory serves me right, you'd just run `make` in the root. (Might need some npm love first)

vogelsgesang commented:
`make` worked almost immediately, but I modified the `var mparser` line so your regexp did not match anymore. Therefore I had to adapt my changes of `lib/messageformat.dev.js` first.

In order to avoid broken tests in the individual commits, I rebased my branch.

eemeli commented:
Thank you @vogelsgesang! I merged this, together with a couple more PEG optimizations & some whitespace cleanup. The previously failing node 0.8 tests were taken out in #88; we now test in node 0.10 & 0.12 and in iojs.

The current custom inlining of the compiled PEG parser will require some thought, and probably replacement.
